File: java_version/bfs/Binary_Tree_Level_Order_Traversal_II.java
package bfs;

import java.util.*;

//use bfs level order traverse == > then reverse
public class Binary_Tree_Level_Order_Traversal_II {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null){
            return result;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        while(!queue.isEmpty()){
            List<Integer> level = new ArrayList<>();
            int size = queue.size();
            for(int i = 0; i < size; i++){
                TreeNode node = queue.poll();
                if(node.left != null){
                    queue.add(node.left);
                }
                if(node.right != null){
                    queue.add(node.right);
                }
                level.add(node.val);
            }
            result.add(level);
        }


        int right = result.size() - 1;
        List<List<Integer>> final_result = new ArrayList<>();

        while(right >= 0){
            final_result.add(result.get(right));
            right -= 1;
        }


        return final_result;
    }
}

File: binary_tree-master/binary-tree-level-order-traversal-ii.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: A tree
    @return: buttom-up level order a list of lists of integer
    """
    def levelOrderBottom(self, root):
        # write your code here
        if root is None:
            return []
        result = []
        queue = [root]
        while queue != [] :
            mid_result = []
            result.append([i.val for i in queue])
            for i in queue:
                if i.left != '#' and i.left is not None:
                    mid_result.append(i.left)
                if i.right != '#' and i.right is not None:
                    mid_result.append(i.right)
            queue = mid_result
        
        result.reverse()
        return result
        
        
